### PR TITLE
Extrapolate stage and expand redoc

### PR DESCRIPTION
**Pull request by mikolak-net**
PR for #23804 . Adds a new stage `pressurize`, and clarifies documentation of `expand`.

--------------
Naming the stage:

This was probably the most difficult portion of the PR. Here are the notes on various possibilities I've come up with:

 - `replay`: as in http://reactivex.io/documentation/operators/replay.html
   - does not imply existence of original element
   - has quite different meaning in semantic contexts of other frameworks
 - `extrapolate`
   - good name
   - but but too phonetically close to `expand` to not be confusing
   - on top of that, `expand` already has an *argument* `extrapolate` (which as far as I understand can't be renamed without breaking binary compatibility, correct me if I'm wrong on this)
 - `project` 
   - not really intuitive
   - already has a different meaning in e.g. functional programming
 - `pressurize`
   - is a completely new term in the reactive context
   - however, very á propos the use case
 - `bounce`: as antonym from debounce e.g. https://monix.io/api/2.3/monix/reactive/observables/ObservableLike.html#debounce(timeout:scala.concurrent.duration.FiniteDuration):Self[A] , http://reactivex.io/documentation/operators/debounce.html
   - usage here does not make sense for anyone knowing the etymology
 - `echo`: analogous to https://monix.io/api/2.3/monix/reactive/observables/ObservableLike.html#echoRepeated(timeout:scala.concurrent.duration.FiniteDuration):Self[A]
    - similar problem as extrapolate (echo / expand)

-------------
Other notes:

0. The binary compatibility check failed, but as I understand in this case (adding new API elements), this is fine, correct?
1. Originally, I started with implementing `pressurize` as a completely new stage, but I remembered that some stages, e.g. `Sink.foreach`, are already written as "compositions" of multiple existing stages, so I went with that approach in the end.
2. I'm not sure if I handled the variance of the stage correctly. The intent was to limit the `extrapolate` function to the original element type, so as to prevent using it as a "poor man's map with bells and whistles".
3. I used `collection.Stream` as the return value of the `extrapolate` function, please let me know if this unorthodox.
4. The tests are mostly copied from `Expand` (since WET seems to be in effect here anyway), and adjusted accordingly.
5. I also applied the Scouts' Rule and cleaned up `expand` - removed the reference to the `seed` argument, and (subjectively) clarified in the tests what `expand` actually does, and replaced references to "publisher" with "upstream" for consistency's sake.
6. I'm not sure whether to include a reference to `pressurize` in the Cookbook. Should I, and, if so, where?

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @mikolak-net,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by raboof** [comment]
OK TO TEST

**Comment by mikolak-net** [comment]
Binary compatibility check failed, as said in the initial notes. To clarify - is there anything actionable for me at this point?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Incomplete sentence

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Ah, I see now why `expand` always passes the element through the function, the type signature is trickier/less useful this way with the lower type bound as it is easy to just widen it to anything. Very unsure how we can get this to work for Java without making the type parameters `Object` across the board as there is no such thing as lower type bounds in Java.

I think this should instead be called call `expandOrEmit` which would highlight the difference between this and the existing one, but, skip the initial as that can be achieved through composition (maybe mention in docs though), the function should not produce `Stream` as that is less generic, has memory leak issues, and limits use cases, instead it should produce an `Iterator` (possibly a custom this-value-and-then-the-rest-iterator of some kind for performance reasons).

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: It should be even more clear that if there is enough elements flowing from upstream the extrapolation is not invoked, and "if you want to always have values pass through the extrapolate function, see [[link to old method]]"

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: We should also have a see link to `expand`

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: We should make it even more clear that extrapolate is always applied, and "if you want elements to flow through unchanged when upstream is keeping up with downstream, use [[link to the new method]]"

File: akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: This will not work for Java (in fact I think we just realized not even the upper bounds really work like we have them). I think we'll have to limit it to `Function[Out, Iterator[Out]]` for Java.

Use `:javap f.q.c.Name` to verify in the sbt or scala console

**Review by johanandren — COMMENTED** [review]
Added some comments.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Naming is hard, sure but we can stick to simply `extrapolate`, since:

> good name

yup

>    but but too phonetically close to expand to not be confusing

Does not matter? I mean, it's a close operation anyway, so that's good.

>    on top of that, expand already has an argument extrapolate (which as far as I understand can't be renamed without breaking binary compatibility, correct me if I'm wrong on this)

Does not really matter I think. Parameter names can be changed in a bin compat way btw, it's doable via `@deprecatedName`, but I don't think there's a need to do this. Pressurize sounds very weird as the only thing about pressure we call such is backpressure, and while this one does relate to it it's a bit of a weird name ;)

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: I would really want to avoid the Stream type, its memoization is not really something we need here, can just emit a strict collection

File: akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala]: good rename

File: akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala]: to progress *independently* of a

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: docs should explain how it differs from the very similar stage :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Yeah and compare them a bit, in both docs

File: akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: Sorry, I'm not sure I understand (especially given the remarks below). What solution do you suggest and what should I verify in the bytecode?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: +1, see later comment.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Gah, I have no idea how I missed that, sorry.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: +1, see later comment.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Yup, mutually referencing the two methods to disambiguate is a good idea here, thanks.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Regarding lower type bounds - not exactly, you can specify wildcard lower type bounds (`<? super Sth>`). It appears it's useless in our case anyway, i.e. the user effectively operates on `Object` anyway.

Re `Stream` - OK, so there's a reason to use that beyond the namespace collision. Could you elaborate on the memory leak issues, by the way? Do you mean the increased overhead as described [here](https://stackoverflow.com/questions/20360161/streamfilter-runs-out-of-memory-for-1-000-000-items), or something else?

Re `Iterator` - any specific concerns you're thinking about using a custom one? The Scala standard library's `Iterator.++` seems to be optimized for efficiency already. 

Re `expandOrEmit` - I'm not partial to names with logical operators for several reasons, the one most relevant here is that they inhibit recall of the semantics (since you have to remember two "subthings"). I'll switch to `extrapolate` on @ktoso 's suggestion (see below) for now.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: In fairness, "backpressure" is probably also a new concept for the majority of Akka Streams newcomers, hence the suggestion. But I won't argue - `extrapolate` is the better choice obviously :).

Re bin compat - what I had in mind is a potential problem with named argument assignments. But on second thought, that should be transformed by compiler into positional arguments anyway - and, from the emitted bytecode, I see that's the case, at least in Scala 2.12.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Acknowledged, changing to `Iterator`, as with `expand`.

File: akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala]: Thanks. I'm thinking about a separate PR that unifies this type of terminology for the entire API.

File: akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala]: Yup, missed that, sorry!

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: +1, see above.

**Comment by mikolak-net** [comment]
Also, I've noticed that `streams-rate.md` references `expand` explicitly, so I'll adjust that once we stabilize the API changes.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Thanks :+

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: > Re bin compat - what I had in mind is a potential problem with named argument assignments. 

As I said, `@deprecatedName` fixes this

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: > Re Stream - OK, so there's a reason to use that beyond the namespace collision. Could you elaborate on the memory leak issues, by the way? Do you mean the increased overhead as described here, or something else?

it memorizes, and that's not something we need here ;) Naming things aside, it's not really a collection that is very useful unless you specifically know you want memoization

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (mikolak-net) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: @johanandren : is something like this what you had in mind as a possible solution to the contravariance problem?

File: akka-stream/src/main/mima-filters/2.5.10.backwards.excludes
Comment (mikolak-net) [inline akka-stream/src/main/mima-filters/2.5.10.backwards.excludes]: I'm assuming adding these ignores is acceptable, given the subject matter of the PR?

Comment (ktoso) [reply akka-stream/src/main/mima-filters/2.5.10.backwards.excludes]: Right, absolutely :)

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Just realized I left this thread dangling in the wind, sorry. Yup, you're right of course that `@deprecatedName` would be a way to handle the concern (thanks BTW for pointing this out, I learned something new).

However, if I understand correctly, you don't see a point in going through the whole deprecation procedure in this case, right?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (mikolak-net) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: One more thing I noticed today - this does not render well within JavaDoc - for the Java API versions, I'm going to move the description portion to the main body of the doc (as suggested by [this related SO answer](https://stackoverflow.com/questions/5011291/usage-of-see-in-javadoc#5014784)), and leave `@see` with just the method names. This is because the only other option would be to use the fragile, "raw" `a` link.

**Comment by mikolak-net** [comment]
I went ahead and added the doc modifications. @ktoso @johanandren please let me know if there are any other outstanding issues that I need to address (esp. [this comment](https://github.com/akka/akka/pull/24431#discussion_r166003773) ).

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Yes, that's what I was thinking we need to do.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: 👍

File: akka-docs/src/main/paradox/stream/stages-overview.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/stages-overview.md]: Isn't this rather, "when there is downstream demand before upstream emitted any element"?

File: akka-docs/src/main/paradox/stream/stages-overview.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/stages-overview.md]: I think the difference from the new `extrapolate` can be made more clear. Maybe something like "Elements are always passed through the function regardless if there is downstream back pressure or not.". Or maybe link to the longer "understanding extrapolate and expand" section?

File: akka-docs/src/main/paradox/stream/stream-rate.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/stream-rate.md]: Same here as above, "For situations where there may be downstream demand before any element is emitted from upstream", it's not really related to "several producers"

File: akka-docs/src/main/paradox/stream/stream-rate.md
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/stream-rate.md]: I think here you should put a more detailed description of the difference between `expand` and `extrapolate`, something like this:

"While `extrapolate` is only applied when there is back pressure, `expand` will always applied to the element going through the stream, this makes some additional things such as transformations and filtering possible.

Here is the previous example with `expand` which makes it somewhat more concise:" 

And then maybe just skip the text below

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Old name, "`extrapolate` does not support", and in some more places below

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: 👍

**Review by johanandren — COMMENTED** [review]
Looks good, added some comments about the docs mostly.

File: akka-docs/src/main/paradox/stream/stages-overview.md
Comment (mikolak-net) [reply akka-docs/src/main/paradox/stream/stages-overview.md]: That's what the last part of the sentence is meant to convey ("and there are no new elements from the upstream").

But since you've missed that, it's probably unclear. I'll rewrite mostly per your suggestion.

File: akka-docs/src/main/paradox/stream/stages-overview.md
Comment (mikolak-net) [reply akka-docs/src/main/paradox/stream/stages-overview.md]: I think a link is the better idea - the examples contained therein will demonstrate the differences better than any "wordy" elaboration.

File: akka-docs/src/main/paradox/stream/stream-rate.md
Comment (mikolak-net) [reply akka-docs/src/main/paradox/stream/stream-rate.md]: You're right, that's only an example scenario, not a description of the general problem case.

File: akka-docs/src/main/paradox/stream/stream-rate.md
Comment (mikolak-net) [reply akka-docs/src/main/paradox/stream/stream-rate.md]: I'd prefer to leave the more verbose explanation where it is - from my observations readers often "zone out" to the explanations in-between when reading snippets, having it at the end will serve to reduce that effect.

I've reworded the subsequent text to refer to the examples more directly.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
Comment (mikolak-net) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Sorry, missed that, searched & replaced.

**Comment by mikolak-net** [comment]
PLS BUILD

**Comment by mikolak-net** [comment]
^it looks like there's a transient build process issue, could someone bump this once that's cleared up?

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
PLS BUILD

**Comment by mikolak-net** [comment]
@ktoso : is there anything I missed, and you need me to address?

**Comment by johanandren** [comment]
Waiting for a second person in the core team to review before we can merge.

**Comment by mikolak-net** [comment]
Yup, I know, that's why I queried `@ktoso`, as I'm assuming that would probably him given he already reviewed the previous iterations, and he might be waiting on something that I've overlooked :).

**Comment by ktoso** [comment]
Very sorry for the delay here, team meeting and travel accumulated a bit. I’ll try to review soon

**Comment by mikolak-net** [comment]
@ktoso : bump :).

**Comment by johanandren** [comment]
Some things you can do to make the merge quicker after that second review: rebase and move mima excludes to a 2.5.11 file as we have released 2.5.11 while this has been waiting (sorry!)

**Comment by mikolak-net** [comment]
@johanandren : thanks for the info, made the changes.

**Review by patriknw — APPROVED** [review]
@mikolak-net LGTM, please rebase to fix the merge conflict (you can squash to one commit if you like), and then we can merge this

**Comment by mikolak-net** [comment]
@patriknw : thanks, rebased (I've left the commits as they are, since as far as I understand the Contributing Guideline they will be squashed anyway, correct?). 

Also, judging from the local build there seems to be a bincompat issue for akka-persistence in the current master (`akka-stream` itself is fine), hope that clears up soon.

**Pull request merged**
